entrypoint.py

```diff
@@ -83,6 +85,51 @@ async def pub_handler(request):
     return web.Response(text="success")
 
 
+# ws = new WebSocket("ws://${api_addr}/api/pub_ws")
+# ws.onopen = () => {
+#     ws.send(JSON.stringify({
```

Comment:
I don't think this comment matches the code.
The first message, in the example, is missing a packet.

You could update the code such that the first message specifies the container+topic, then follow up messages only need the packet. This would also remove the need for the global `publishers`.

Reply:
It would be nice to be able to form the connection without sending a packet. I can imagine use-cases for us where we'll want to make a connection to an outside process before it actively wants to send a message.

On the second point, could you elaborate? Are you talking about using a closure somehow?

Reply:
Here's what I'm thinking:

```py
# ws = new WebSocket("ws://${api_addr}/api/pub_ws")
# ws.onopen = () => {
#     ws.send(JSON.stringify({
#         container: "...",
#         topic: "...",
#     }))
# }
#
# then:
#
# ws.send(JSON.stringify({
#         packet: {
#             headers: [
#                 ["key", "val"],
#                 ...
#             ],
#             payload: window.btoa("..."),
#         },
#     }))
async def pub_handler(request):
    ws = web.WebSocketResponse()
    await ws.prepare(request)

    publisher = None

    async for msg in ws:
        if msg.type != WSMsgType.TEXT:
            break

        cmd = json.loads(msg.data)

        if publisher is None:
          tm = a0.TopicManager(container=cmd["container"])
          publisher = a0.Publisher(tm.publisher_topic(cmd["topic"]))
          continue

        publisher.pub(
            a0.Packet(
                cmd["packet"]["headers"],
                base64.b64decode(cmd["packet"]["payload"])
            )
        )
```

Reply:
I see. I didn't understand how the `async` worked on this `for` loop. Thanks! All updated.